### PR TITLE
Flag to check if authorizations to collect HealthKit data are already present

### PR DESCRIPTION
**Pull request by philippzagar**
# Flag to check if authorizations to collect HealthKit data are already present

## :recycle: Current situation & Problem
This pull request addresses an identified gap in the functionality provided by the `SpeziHealthKit` integration. 
Currently, there isn't an accessible API that allows application developers to verify whether the necessary HealthKit permissions, required to collect specified health data samples, have already been authorized by the user. This may result in **redundant authorization requests** in which applications re-request those permissions.
A common example of this problem can be observed in the current implementation of the [SpeziTemplateApplication](https://github.com/StanfordSpezi/SpeziTemplateApplication). It erroneously prompts an onboarding step that requests HealthKit permissions from the user, even if such permissions have previously been granted (e.g. by forcing the onboarding flow via the `--showOnboarding` flag in the schema).

This issue is addressed in #6.

## :bulb: Proposed solution
The PR aims to enhance the `HealthKit` module by implementing a check for existing HealthKit authorizations, thereby eliminating unnecessary re-authorization requests and improving the user experience. This is achieved by providing the `public func checkAuthorizations() async -> Bool` function within the `HealthKit` module that compares the necessary HealthKit authorizations with the ones already given by HealthKit (via querying the `HKHealthStore`). If all requested authorizations are already given, the function returns `true`, otherwise `false`. 

Application developers can then use this function to e.g. skip the HealthKit authorization onboarding step, as exemplified in this brief code example:
```swift
struct SomeView1: View {
    @EnvironmentObject var healthKitDataSource: HealthKit<FHIR>

    var body: some View {
         SomeView2(
              action: {
                  if await healthKitDataSource.checkAuthorizations() {
                        // Skip authorization step
                  } else {
                        // Show authorization step
                  }
              }
         )
    }
}
```

## :gear: Release Notes
- Extend the public API of the `HealthKit` module by the function `checkAuthorizations()` to check if all necessary HealthKit authorizations are already granted

## :heavy_plus_sign: Additional Information
This is my first PR within this project, feel free to give me lots of feedback, especially regarding the setup of the test cases :)

### Related PRs
[SpeziTemplateApplication PR](https://github.com/StanfordSpezi/SpeziTemplateApplication/pull/26)

### Testing
Unit test cases as well as small UITests are included that test all scenarios of the introduced `checkAuthorizations()` function

### Reviewer Nudging
The actual code changes aren't too big, take a closer look at the setup of the test cases.

### Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [X] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Comment by philippzagar** [comment]
It seems like the infrastructure of the test cases currently has some runner(?) issues. Locally, both unit tests as well as UITests are successful.

**Comment by vishnuravi** [comment]
Hi @philippzagar this is a great idea! Thank you for implementing it! One question I have is, how do we handle the situation where the user has provided permissions, but then revokes those permissions from within the Apple Health app?

**Comment by philippzagar** [comment]
Hi @vishnuravi,
thanks for the feedback! 🚀

You definitely raise a good point here. My reasoning for the current implementation is the structure of the `askForAuthorization()` function which relies on writing the authorized data sources to `UserDefaults`. Therefore, I built on top of this concept initially.

However, I did some reengineering of this PR that eliminates this primitive data storage altogether (so within `UserDefaults`). The `authorized` property (now a function named `checkAuthorizations()` as `async`) as well as the `askForAuthorization()` function now directly query the `HKHealthStore` to check if the respective HealthKit authorizations are already given. This addresses your concern so that the authorizations within the app are always up to date, regardless if they are revoked at a later point in time via the Apple Health settings.

Let me know what you think about the current implementation. I also updated the unit test cases (using a Spy for the `HKHealthStore`) as well as the UI tests.

I'll create a PR for utilizing this property within the [SpeziTemplateApplication](https://github.com/StanfordSpezi/SpeziTemplateApplication) soon

**Comment by philippzagar** [comment]
Another important question would be: How should we go forward with other usages of `UserDefaults.Keys.healthKitRequestedSampleTypes`? Is that something we'd like to keep or get rid of it altogether by querying the `HKHealthStore` directly (keep in mind, that's an `async` operation)

**Review by PSchmiedmayer — CHANGES_REQUESTED** [review]
@philippzagar Thank you so much for the changes and improvements in this PR! It looks great!

I only had a few smaller change request that I documented in the issues. Thank you for exploring the different avenues together with @vishnuravi. Both of you came to the right conclusion, I agree that the local storage mechanism is the best way we can currently support that functionality. We can't really see if the user has denied access after the fact and will also not be able to detect that anyways.

One thing would be cool to investigate: Is it even possible to show the permissions sheet again if the user has denied access after the fact? If not, then this is a 100% confirmation that we are doing the right thing 👍

I have also addressed the CI issue in https://github.com/StanfordSpezi/.github/releases/tag/v2.0.15, this should propagate to this PR, I will commit a minimal change of the SwiftLint configuration to trigger the CI 👍

**Comment by PSchmiedmayer** [comment]
@philippzagar While the testing behaviour is a bit flakey here, the following latest release might help a bit with the stability here, I would suggest updating the UI testing app to the releases:
- https://github.com/StanfordBDHG/XCTestExtensions/releases/tag/0.4.6
- https://github.com/StanfordBDHG/XCTHealthKit/releases/tag/0.3.4

File: Tests/UITests/TestApp/HealthKitTestsView.swift [hunk 1]
```diff
@@ -48,6 +50,8 @@ struct HealthKitTestsView: View {
     private func askForAuthorization() {
         Task {
             try await healthKitComponent.askForAuthorization()
+            // Required as authorization button isn't rerendered otherwise
+            self.healthKitComponent.objectWillChange.send()
```
[the remark this answers is not in the document]
Comment (philippzagar) [reply Tests/UITests/TestApp/HealthKitTestsView.swift:54]: That makes sense, thanks for the hint!

**Comment by philippzagar** [comment]
@PSchmiedmayer Thanks for the in-depth feedback! 🚀

Sadly, the UI Test behavior is quite flakey. I already upgraded the testing dependencies to the most up-to-date versions.
Locally, the UI tests complete successfully every time, on the GitHub self-hosted runner it fails most of the time. Maybe we should take a closer look into that 😅. 

Regarding detecting HealthKit authorizations after the fact: As outlined in this [StackOverflow thread](https://stackoverflow.com/questions/29076655/healthkit-authorisation-status-is-always-1/29128231#29128231) as well as the [Apple docs](https://developer.apple.com/documentation/healthkit/hkauthorizationstatus), HealthKit doesn't allow AT ALL to determine if the user gives read authorizations or denies them, even when directly prompting the user with the HealthKit authorization view the first time. Said that, determining whether read authorizations are revoked after the fact is impossible, at least from the native HealthKit side.
However, this is not true for write (share) permissions. Here, one is able to differentiate between given / revoked write authorizations. 

Directly from the [Apple docs](https://developer.apple.com/documentation/healthkit/hkauthorizationstatus):
> To help maintain the privacy of sensitive health data, HealthKit does not tell you when the user denies your app permission to query data. Instead, it simply appears as if HealthKit does not have any data matching your query. Your app will receive only the data that it has written to HealthKit. Data from other sources remains hidden from your app. For more information on privacy in HealthKit, see HealthKit.


A bit of a hacky workaround to determine if a user revoked authorizations after the fact: When the user grants authorizations for the first time, we could store which data types actually have data to collect (remember: we cannot differentiate if a user gives no authorizations or there just isn't any data available for this type). Based on this data, we could query the `HKHealthStore` every time one opens the app and check if we're still able to collect at least one data sample for each type that we were able to collect directly after the initial authorization. If this data set matches, we know that there weren't any revokes in the meantime. If it doesn't, we can assume that authorizations were revoked. 
However, note that this is quite a hacky workaround that doesn't guarantee correctness (e.g. health data of a specific type could be deleted by the user in the meantime but the authorizations are still in place). 
How would you like to proceed here @PSchmiedmayer? From my perspective, I would suggest not going down this route. I think revoking health authorizations for a specific app can also be considered to be an edge case.

**Comment by vishnuravi** [comment]
@philippzagar thank you for taking the time to look into this issue in detail and come up with possible solutions. Although the workaround you have proposed is very clever, I agree that it makes more sense to stick to your original approach. The situation in which a user would revoke permissions from within Apple Health after the fact is likely not common enough to be worth implementing a hacky workaround. Do you agree @PSchmiedmayer?

**Comment by philippzagar** [comment]
> The situation in which a user would revoke permissions from within Apple Health after the fact is likely not common enough to be worth implementing a hacky workaround.

As stated in my comment, I couldn't agree more, this is definitely an edge case that will affect a very small percentage of users. 
However, I'm still convinced that Spezi needs to give some sort of indication that the user ("probably") doesn't grant the required authorizations. Any other ideas on how to achieve this @PSchmiedmayer? I already googled a bit and it seems like a dead end to me, except for the hacky workaround I propose above to determine possible revokes **after the fact** (so when authorizations were granted at least for a short amount of time), not directly after the initial HealthKit authorization prompt.

**Comment by philippzagar** [comment]
@PSchmiedmayer Regarding the flakey UI tests:
Here a representative error log, seems like relaunching the app is causing some problems

```
2023-07-04 07:52:22.214 xcodebuild[42592:1954547] [MT] IDELaunchReport: b9a73f618c063800:b9a73f618c063800: Finished with error: Simulator device returned an error for the requested operation.
Domain: FBSOpenApplicationServiceErrorDomain
Code: 4
User Info: {
    BSErrorCodeDescription = InvalidRequest;
    FBSOpenApplicationRequestID = 0xcfe3;
    IDERunOperationFailingWorker = IDELaunchiPhoneSimulatorLauncher;
    SimCallingSelector = "launchApplicationWithID:options:pid:error:";
}
--
The request to open "edu.stanford.spezi.healthkit.testapp" failed.
Domain: FBSOpenApplicationServiceErrorDomain
Code: 4
User Info: {
    BSErrorCodeDescription = InvalidRequest;
    FBSOpenApplicationRequestID = 0xcfe3;
}
--
The operation couldn’t be completed. Application info provider (FBSApplicationLibrary) returned nil for "edu.stanford.spezi.healthkit.testapp"
Domain: FBSOpenApplicationErrorDomain
Code: 4
Failure Reason: Application info provider (FBSApplicationLibrary) returned nil for "edu.stanford.spezi.healthkit.testapp"
User Info: {
    BSErrorCodeDescription = NotFound;
}
--
```

**Comment by PSchmiedmayer** [comment]
@philippzagar @vishnuravi I agree, let's stick with the current approach and assume that the user provided adequate permissions and if not, it is their choice and we have to respect that anyways. We should not and can not enforce that they provide permissions and therefore just need to be sure that we asked them at least once. We do that with the current approach and I think this is sufficient; I approve the current approach.

In case anybody would like to implement something more custom they can always go back to the comments in this PR and refer to some rather complicated workaround.

**Comment by PSchmiedmayer** [comment]
Regarding the CI: I will need to take a closer look at that. One thing that would be great if you can investigate that @philippzagar: You can download the result bundle of the build from the results page of the GitHub Action: https://github.com/StanfordSpezi/SpeziHealthKit/actions/runs/5455556933. You will have to download the bundle and open it in Xcode 15 to see a video recording of the UI test. 

This might provide us more insights into the exact mechanism that failed. In case this is something in XCTestExtensions or XCTHealthKit we might be able to address this with further timeouts or additional checks. Sometimes just adding an additional `sleep(...)` call at the right time helps when dealing with the reduced performance of the runners.

**Comment by philippzagar** [comment]
@PSchmiedmayer After some painful debugging hours, I finally found the reason for the UI Test failure. It's a very weird behaviour within the Apple Health app in which the UI Tests detect two buttons named "High Heart Rate", one in the foreground sheet used to add health data (which we also intend to detect) and one behind the sheet displaying the current health records of high heart rates (which we don't intend to detect).

**Intended** "High Heart Rate" button
image

**Accidentally** detected "High Heart Rate" button
image


I couldn't reproduce the problem locally until I upgraded to Xcode 15, then I managed to reproduce it once. I guess it's an iOS 17 thing that wasn't yet noticed within the test cases of this repo (as there wasn't a PR in quite some time).

As I'm not yet a member of the `StanfordBDHG` orga, I published an adjusted version of the `XCTHealthKit` in `[my personal account](https://github.com/philippzagar/XCTHealthKit/pull/1/files)`. Feel free to check out the changes there.
As already discussed via DM, @PSchmiedmayer please add me to the `StanfordBDHG` orga so that I can also create a proper PR there :)

**Comment by PSchmiedmayer** [comment]
@philippzagar Thank you for the further investigation, greatly appreciated! I have merged the PR over at XCTHealthKit and tagged a new release: https://github.com/StanfordBDHG/XCTHealthKit/releases/tag/0.3.5

File: Tests/UITests/UITests.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -634,15 +634,15 @@
...
-		2F85828029E7784C0021D637 /* XCRemoteSwiftPackageReference "XCTHealthKit" */ = {
+		97A7D7162A54D69700D85923 /* XCRemoteSwiftPackageReference "XCTHealthKit" */ = {
 			isa = XCRemoteSwiftPackageReference;
-			repositoryURL = "https://github.com/StanfordBDHG/XCTHealthKit";
+			repositoryURL = "https://github.com/philippzagar/XCTHealthKit";
 			requirement = {
-				kind = upToNextMinorVersion;
-				minimumVersion = 0.3.3;
+				branch = main;
+				kind = branch;
```
Comment (PSchmiedmayer) [inline Tests/UITests/UITests.xcodeproj/project.pbxproj:645]: Short comment to remember that we chance that back to the upstream version before we tag the release.

**Review by PSchmiedmayer — APPROVED** [review]
Looks great @philippzagar, thank you for the updates! 👍

**Pull request merged**